video: Patch SDMA & capture sample to arbitrate height and width

zacck opened the pull request:
The  video capture sample does not work when using a combination of the FRDM-MCXN236  and the OV5640 camera and this is due to the sensor and the underlying interface having different defaults. This patch sets the interface resolution to that selected by the sample which in turn is set in the CONFIG.

JarmouniA reviewed (commented):
Please split this into 2 commits: driver changes, sample changes.
Also, explain why the driver change is needed in the commit message.

zacck committed:
drivers: video: mcux_sdma: Reconfigure the source when getting format

Reconfigure the sensor format with the only supported format of the
sdma when getting format

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck commented:
> Please split this into 2 commits: driver changes, sample changes. Also, explain why the driver change is needed in the commit message.

Thank you, this is done.

josuah commented on drivers/video/video_mcux_smartdma.c:
```diff
@@ -267,7 +267,12 @@ static int nxp_video_sdma_get_format(const struct device *dev, struct video_form
 	if ((fmt->pixelformat != fmts[0].pixelformat) ||
 	    (fmt->width != fmts[0].width_min) ||
 	    (fmt->height != fmts[0].height_min)) {
-		return -ENOTSUP;
+		ret = video_set_format(config->sensor_dev, fmt);
+
+		if (ret < 0) {
+			LOG_ERR("Sensor device does not support RGB565");
+			return ret;
+		}
```
That indeed fixes the bug encountered by the capture sample.

The fix for this is ultimately in the `capture` sample IMHO, but for the sake of a workaround that stll makes sense.

Another compromise in the meantime is to configure the default format in `_init()` and make this call always return the default supported format without asking the source?

josuah replied:
If you wanted to make sure that `fmt = RGB565 320x240`, you would also need to use `fmts[0]` instead of `fmt`: as this is:

```c
if (/* format is not what is expected */) {
    /* use the incorrect fmt */
}
```

So that would need a `tmp = { /* correct fmt */ }` if wanting to go that path (but +1 for simply moving that in `init()`).

zacck replied:
Ah I think I see your way and I prefer it, so at unit we can set format to fmts[0], I think that’s better because this is the default format  for SDMA and if anything deviates it should be flagged rather than coercing an error into the right thing. Will push this change shortly.

zacck replied:
@ngphibang  makes sense shall do this now.

zacck replied:
@ngphibang  I just did this and there is now just one commit I didn't move the work to _init as if it is initialized before the sensor then it won't work cc @josuah

ngphibang reviewed (dismissed):
Could you confirm again this sole commit fixes the issue? I don't have HW to test.

zacck committed:
samples: video: capture: Configure frame width and heght for mcxn236

SDMA as the video interface on this board requires
this size frame to be set for the sample to work

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

ngphibang reviewed (changes requested):
Although the 2nd commit is not really needed for the capture sample, configuring the frame width & height is still OK to ensure that if, for some reason, the sample application does not get the format before setting the format, it's still working.

BTW, could you change the commits title, to something like :

```
drivers: video: mcux_sdma: Reconfigure the source when getting format
Reconfigure the sensor format with the only supported format of the sdma when getting format
```

`samples: video: capture: Configure frame width and heght for mcxn236`

Thanks !

ngphibang reviewed (approved):
The 2nd commit message needs a bit rewording but not a big problem to me. Thanks.

josuah reviewed (approved):
Thanks for the commits/testing/bugfix!
This will help working with more sensors.
